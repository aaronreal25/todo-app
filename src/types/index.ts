export type TaskStatus = 'PENDING' | 'IN_PROGRESS' | 'COMPLETED';

export interface Task {
  id: string;
  title: string;
  description: string;
  status: TaskStatus;
  dueDate: string;
  createdAt: string;
}

export interface TaskFilter {
  status: TaskStatus | 'ALL';
  searchQuery: string;
}

export interface TaskState {
  tasks: Task[];
  filter: TaskFilter;
  isLoading: boolean;
  error: string | null;
}

export type TaskAction =
  | { type: 'SET_TASKS'; payload: Task[] }
  | { type: 'ADD_TASK'; payload: Task }
  | { type: 'UPDATE_TASK'; payload: Task }
  | { type: 'DELETE_TASK'; payload: string }
  | { type: 'SET_FILTER'; payload: Partial<TaskFilter> }
  | { type: 'SET_LOADING'; payload: boolean }
  | { type: 'SET_ERROR'; payload: string | null };