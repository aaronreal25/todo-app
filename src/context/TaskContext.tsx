'use client';

import React, { createContext, useContext, useReducer, useCallback } from 'react';
import { Task, TaskStatus } from '@/types';

interface TaskState {
  tasks: Task[];
  isLoading: boolean;
  error: string | null;
}

type TaskAction =
  | { type: 'ADD_TASK'; payload: Task }
  | { type: 'UPDATE_TASK'; payload: Task }
  | { type: 'DELETE_TASK'; payload: string }
  | { type: 'SET_LOADING'; payload: boolean }
  | { type: 'SET_ERROR'; payload: string | null };

const initialState: TaskState = {
  tasks: [],
  isLoading: false,
  error: null,
};

function taskReducer(state: TaskState, action: TaskAction): TaskState {
  switch (action.type) {
    case 'ADD_TASK':
      return {
        ...state,
        tasks: [...state.tasks, action.payload],
        isLoading: false,
        error: null,
      };
    
    case 'UPDATE_TASK':
      return {
        ...state,
        tasks: state.tasks.map((task) =>
          task.id === action.payload.id ? action.payload : task
        ),
      };
    
    case 'DELETE_TASK':
      return {
        ...state,
        tasks: state.tasks.filter((task) => task.id !== action.payload),
      };
    
    case 'SET_LOADING':
      return {
        ...state,
        isLoading: action.payload,
      };
    
    case 'SET_ERROR':
      return {
        ...state,
        error: action.payload,
      };
    
    default:
      return state;
  }
}

// Export the context
export const TaskContext = createContext<{
  state: TaskState;
  addTask: (task: Omit<Task, 'id' | 'createdAt'>) => void;
  updateTask: (task: Task) => void;
  deleteTask: (id: string) => void;
} | null>(null);

export const TaskProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [state, dispatch] = useReducer(taskReducer, initialState);

  const addTask = useCallback((taskData: Omit<Task, 'id' | 'createdAt'>) => {
    const newTask: Task = {
      ...taskData,
      id: crypto.randomUUID(),
      createdAt: new Date().toISOString(),
    };
    dispatch({ type: 'ADD_TASK', payload: newTask });
  }, []);

  const updateTask = useCallback((task: Task) => {
    dispatch({ type: 'UPDATE_TASK', payload: task });
  }, []);

  const deleteTask = useCallback((id: string) => {
    dispatch({ type: 'DELETE_TASK', payload: id });
  }, []);

  return (
    <TaskContext.Provider 
      value={{ 
        state,
        addTask,
        updateTask,
        deleteTask,
      }}
    >
      {children}
    </TaskContext.Provider>
  );
};

export const useTaskContext = () => {
  const context = useContext(TaskContext);
  if (!context) {
    throw new Error('useTaskContext must be used within a TaskProvider');
  }
  return context;
};
