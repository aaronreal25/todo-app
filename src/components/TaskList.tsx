'use client';

import React, { useState, useMemo } from 'react';
import { FixedSizeList } from 'react-window';
import TaskItem from './TaskItem';
import { Task, TaskStatus } from '@/types';
import { useTaskContext } from '@/context/TaskContext';

interface TaskListProps {
  onEdit: (task: Task) => void;
}

interface Filter {
  status: TaskStatus | 'ALL';
  searchQuery: string;
}

const ITEM_HEIGHT = 200; // Height of each task item in pixels
const WINDOW_HEIGHT = 800; // Height of the virtualized window

const TaskList: React.FC<TaskListProps> = React.memo(({ onEdit }) => {
  const { state, updateTask, deleteTask } = useTaskContext();
  const [filter, setFilter] = useState<Filter>({
    status: 'ALL',
    searchQuery: '',
  });

  const filteredTasks = useMemo(() => {
    return state.tasks.filter(task => {
      const matchesStatus = filter.status === 'ALL' || task.status === filter.status;
      const matchesSearch = task.title.toLowerCase().includes(filter.searchQuery.toLowerCase());
      return matchesStatus && matchesSearch;
    });
  }, [state.tasks, filter]);

  const Row = React.memo(({ index, style }: { index: number; style: React.CSSProperties }) => {
    const task = filteredTasks[index];
    return (
      <div style={style}>
        <TaskItem
          key={task.id}
          task={task}
          onStatusChange={(id, status) => {
            updateTask({ ...task, status });
          }}
          onEdit={onEdit}
          onDelete={deleteTask}
        />
      </div>
    );
  });
  Row.displayName = 'TaskRow';

  return (
    <div className="task-list-container">
      <div className="task-list-filters mb-6 flex flex-col sm:flex-row gap-4">
        <div className="flex-1">
          <label htmlFor="task-search" className="sr-only">Search tasks</label>
          <input
            id="task-search"
            type="text"
            placeholder="Search tasks..."
            value={filter.searchQuery}
            onChange={(e) => setFilter(prev => ({ ...prev, searchQuery: e.target.value }))}
            className="task-search-input w-full px-4 py-2 bg-gray-800 border border-gray-700 rounded-lg text-white placeholder-gray-400"
          />
        </div>
        
        <div>
          <label htmlFor="status-filter" className="sr-only">Filter by status</label>
          <select
            id="status-filter"
            value={filter.status}
            onChange={(e) => setFilter(prev => ({ ...prev, status: e.target.value as TaskStatus | 'ALL' }))}
            className="task-status-filter px-4 py-2 bg-gray-800 border border-gray-700 rounded-lg text-white"
            aria-label="Filter by status"
          >
            <option value="ALL">All Tasks</option>
            <option value="PENDING">Pending</option>
            <option value="IN_PROGRESS">In Progress</option>
            <option value="COMPLETED">Completed</option>
          </select>
        </div>
      </div>

      {state.tasks.length === 0 ? (
        <div className="task-list-empty text-center py-8 text-gray-400">
          No tasks yet. Add your first task!
        </div>
      ) : filteredTasks.length === 0 ? (
        <div className="task-list-empty text-center py-8 text-gray-400">
          No tasks match your filters.
        </div>
      ) : (
        <FixedSizeList
          height={WINDOW_HEIGHT}
          width="100%"
          itemCount={filteredTasks.length}
          itemSize={ITEM_HEIGHT}
        >
          {Row}
        </FixedSizeList>
      )}
    </div>
  );
});

TaskList.displayName = 'TaskList';

export default TaskList;
