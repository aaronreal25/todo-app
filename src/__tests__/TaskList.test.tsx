import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import TaskList from '@/components/TaskList';
import { TaskContext } from '@/context/TaskContext';
import { Task, TaskStatus } from '@/types';

describe('TaskList', () => {
  const mockOnEdit = jest.fn();

  const mockTasks: Task[] = [
    {
      id: '1',
      title: 'Task 1',
      description: 'Description 1',
      status: 'PENDING' as TaskStatus,
      dueDate: '2024-12-31',
      createdAt: '2024-01-01'
    },
    {
      id: '2',
      title: 'Task 2',
      description: 'Description 2',
      status: 'IN_PROGRESS' as TaskStatus,
      dueDate: '2024-12-31',
      createdAt: '2024-01-01'
    },
    {
      id: '3',
      title: 'Completed Task',
      description: 'Description 3',
      status: 'COMPLETED' as TaskStatus,
      dueDate: '2024-12-31',
      createdAt: '2024-01-01'
    }
  ];

  const mockContextValue = {
    state: {
      tasks: mockTasks,
      isLoading: false,
      error: null
    },
    addTask: jest.fn(),
    updateTask: jest.fn(),
    deleteTask: jest.fn()
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('shows empty state when no tasks', () => {
    const emptyContextValue = {
      ...mockContextValue,
      state: { ...mockContextValue.state, tasks: [] }
    };

    render(
      <TaskContext.Provider value={emptyContextValue}>
        <TaskList onEdit={mockOnEdit} />
      </TaskContext.Provider>
    );

    expect(screen.getByText(/no tasks yet/i)).toBeInTheDocument();
  });

  it('shows no matches message when filter returns no results', () => {
    render(
      <TaskContext.Provider value={mockContextValue}>
        <TaskList onEdit={mockOnEdit} />
      </TaskContext.Provider>
    );

    // Search for non-existent task
    const searchInput = screen.getByPlaceholderText(/search tasks/i);
    fireEvent.change(searchInput, { target: { value: 'nonexistent task' } });

    expect(screen.getByText(/no tasks match your filters/i)).toBeInTheDocument();
  });

  it('filters tasks by status correctly', () => {
    render(
      <TaskContext.Provider value={mockContextValue}>
        <TaskList onEdit={mockOnEdit} />
      </TaskContext.Provider>
    );

    const statusFilter = screen.getByLabelText(/filter by status/i);

    // Test COMPLETED filter
    fireEvent.change(statusFilter, { target: { value: 'COMPLETED' } });
    expect(screen.getByText('Completed Task')).toBeInTheDocument();
    expect(screen.queryByText('Task 1')).not.toBeInTheDocument();
    expect(screen.queryByText('Task 2')).not.toBeInTheDocument();

    // Test PENDING filter
    fireEvent.change(statusFilter, { target: { value: 'PENDING' } });
    expect(screen.getByText('Task 1')).toBeInTheDocument();
    expect(screen.queryByText('Task 2')).not.toBeInTheDocument();
    expect(screen.queryByText('Completed Task')).not.toBeInTheDocument();
  });

  // Snapshot tests
  describe('Snapshots', () => {
    it('matches snapshot with multiple tasks', () => {
      const { container } = render(
        <TaskContext.Provider value={mockContextValue}>
          <TaskList onEdit={mockOnEdit} />
        </TaskContext.Provider>
      );
      expect(container).toMatchSnapshot();
    });

    it('matches snapshot with no tasks', () => {
      const emptyContextValue = {
        ...mockContextValue,
        state: { ...mockContextValue.state, tasks: [] }
      };

      const { container } = render(
        <TaskContext.Provider value={emptyContextValue}>
          <TaskList onEdit={mockOnEdit} />
        </TaskContext.Provider>
      );
      expect(container).toMatchSnapshot();
    });

    it('matches snapshot with filtered tasks', () => {
      const { container } = render(
        <TaskContext.Provider value={mockContextValue}>
          <TaskList onEdit={mockOnEdit} />
        </TaskContext.Provider>
      );

      const statusFilter = screen.getByLabelText(/filter by status/i);
      fireEvent.change(statusFilter, { target: { value: 'COMPLETED' } });

      expect(container).toMatchSnapshot();
    });
  });
});