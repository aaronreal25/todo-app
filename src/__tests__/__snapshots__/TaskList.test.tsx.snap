// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskList Snapshots matches snapshot with filtered tasks 1`] = `
<div>
  <div
    class="task-list-container"
  >
    <div
      class="task-list-filters mb-6 flex flex-col sm:flex-row gap-4"
    >
      <div
        class="flex-1"
      >
        <label
          class="sr-only"
          for="task-search"
        >
          Search tasks
        </label>
        <input
          class="task-search-input w-full px-4 py-2 bg-gray-800 border border-gray-700 rounded-lg text-white placeholder-gray-400"
          id="task-search"
          placeholder="Search tasks..."
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          class="sr-only"
          for="status-filter"
        >
          Filter by status
        </label>
        <select
          aria-label="Filter by status"
          class="task-status-filter px-4 py-2 bg-gray-800 border border-gray-700 rounded-lg text-white"
          id="status-filter"
        >
          <option
            value="ALL"
          >
            All Tasks
          </option>
          <option
            value="PENDING"
          >
            Pending
          </option>
          <option
            value="IN_PROGRESS"
          >
            In Progress
          </option>
          <option
            value="COMPLETED"
          >
            Completed
          </option>
        </select>
      </div>
    </div>
    <div
      style="position: relative; height: 800px; width: 100%; overflow: auto; will-change: transform; direction: ltr;"
    >
      <div
        style="height: 200px; width: 100%;"
      >
        <div
          style="position: absolute; left: 0px; top: 0px; height: 200px; width: 100%;"
        >
          <div
            class="task-item p-4 mb-4 border-2 rounded-lg bg-gray-800 border-green-500"
            data-testid="task-item"
          >
            <div
              class="task-item-header flex justify-between items-center mb-2"
            >
              <h3
                class="task-item-title text-lg font-semibold text-white"
              >
                Completed Task
              </h3>
              <span
                aria-label="completed"
                class="task-item-status px-2 py-1 rounded bg-green-900 text-green-100"
                role="status"
              >
                COMPLETED
              </span>
            </div>
            <p
              class="task-item-description text-gray-300 mb-2"
            >
              Description 3
            </p>
            <p
              class="task-item-due-date text-sm text-gray-400"
            >
              Due: 
              Dec 31, 2024
            </p>
            <div
              class="task-item-controls flex justify-between mt-4"
            >
              <select
                class="task-status-select p-2 border rounded bg-gray-700 text-white border-gray-600 focus:border-blue-500 focus:ring-blue-500"
              >
                <option
                  value="PENDING"
                >
                  Pending
                </option>
                <option
                  value="IN_PROGRESS"
                >
                  In Progress
                </option>
                <option
                  value="COMPLETED"
                >
                  Completed
                </option>
              </select>
              <div
                class="task-item-buttons space-x-2"
              >
                <button
                  class="task-edit-button px-3 py-1 bg-blue-600 text-white rounded hover:bg-blue-700"
                >
                  Edit
                </button>
                <button
                  class="task-delete-button px-3 py-1 bg-red-600 text-white rounded hover:bg-red-600"
                >
                  Delete
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TaskList Snapshots matches snapshot with multiple tasks 1`] = `
<div>
  <div
    class="task-list-container"
  >
    <div
      class="task-list-filters mb-6 flex flex-col sm:flex-row gap-4"
    >
      <div
        class="flex-1"
      >
        <label
          class="sr-only"
          for="task-search"
        >
          Search tasks
        </label>
        <input
          class="task-search-input w-full px-4 py-2 bg-gray-800 border border-gray-700 rounded-lg text-white placeholder-gray-400"
          id="task-search"
          placeholder="Search tasks..."
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          class="sr-only"
          for="status-filter"
        >
          Filter by status
        </label>
        <select
          aria-label="Filter by status"
          class="task-status-filter px-4 py-2 bg-gray-800 border border-gray-700 rounded-lg text-white"
          id="status-filter"
        >
          <option
            value="ALL"
          >
            All Tasks
          </option>
          <option
            value="PENDING"
          >
            Pending
          </option>
          <option
            value="IN_PROGRESS"
          >
            In Progress
          </option>
          <option
            value="COMPLETED"
          >
            Completed
          </option>
        </select>
      </div>
    </div>
    <div
      style="position: relative; height: 800px; width: 100%; overflow: auto; will-change: transform; direction: ltr;"
    >
      <div
        style="height: 600px; width: 100%;"
      >
        <div
          style="position: absolute; left: 0px; top: 0px; height: 200px; width: 100%;"
        >
          <div
            class="task-item p-4 mb-4 border-2 rounded-lg bg-gray-800 border-yellow-500"
            data-testid="task-item"
          >
            <div
              class="task-item-header flex justify-between items-center mb-2"
            >
              <h3
                class="task-item-title text-lg font-semibold text-white"
              >
                Task 1
              </h3>
              <span
                aria-label="pending"
                class="task-item-status px-2 py-1 rounded bg-yellow-900 text-yellow-100"
                role="status"
              >
                PENDING
              </span>
            </div>
            <p
              class="task-item-description text-gray-300 mb-2"
            >
              Description 1
            </p>
            <p
              class="task-item-due-date text-sm text-gray-400"
            >
              Due: 
              Dec 31, 2024
            </p>
            <div
              class="task-item-controls flex justify-between mt-4"
            >
              <select
                class="task-status-select p-2 border rounded bg-gray-700 text-white border-gray-600 focus:border-blue-500 focus:ring-blue-500"
              >
                <option
                  value="PENDING"
                >
                  Pending
                </option>
                <option
                  value="IN_PROGRESS"
                >
                  In Progress
                </option>
                <option
                  value="COMPLETED"
                >
                  Completed
                </option>
              </select>
              <div
                class="task-item-buttons space-x-2"
              >
                <button
                  class="task-edit-button px-3 py-1 bg-blue-600 text-white rounded hover:bg-blue-700"
                >
                  Edit
                </button>
                <button
                  class="task-delete-button px-3 py-1 bg-red-600 text-white rounded hover:bg-red-600"
                >
                  Delete
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          style="position: absolute; left: 0px; top: 200px; height: 200px; width: 100%;"
        >
          <div
            class="task-item p-4 mb-4 border-2 rounded-lg bg-gray-800 border-blue-500"
            data-testid="task-item"
          >
            <div
              class="task-item-header flex justify-between items-center mb-2"
            >
              <h3
                class="task-item-title text-lg font-semibold text-white"
              >
                Task 2
              </h3>
              <span
                aria-label="in_progress"
                class="task-item-status px-2 py-1 rounded bg-blue-900 text-blue-100"
                role="status"
              >
                IN PROGRESS
              </span>
            </div>
            <p
              class="task-item-description text-gray-300 mb-2"
            >
              Description 2
            </p>
            <p
              class="task-item-due-date text-sm text-gray-400"
            >
              Due: 
              Dec 31, 2024
            </p>
            <div
              class="task-item-controls flex justify-between mt-4"
            >
              <select
                class="task-status-select p-2 border rounded bg-gray-700 text-white border-gray-600 focus:border-blue-500 focus:ring-blue-500"
              >
                <option
                  value="PENDING"
                >
                  Pending
                </option>
                <option
                  value="IN_PROGRESS"
                >
                  In Progress
                </option>
                <option
                  value="COMPLETED"
                >
                  Completed
                </option>
              </select>
              <div
                class="task-item-buttons space-x-2"
              >
                <button
                  class="task-edit-button px-3 py-1 bg-blue-600 text-white rounded hover:bg-blue-700"
                >
                  Edit
                </button>
                <button
                  class="task-delete-button px-3 py-1 bg-red-600 text-white rounded hover:bg-red-600"
                >
                  Delete
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          style="position: absolute; left: 0px; top: 400px; height: 200px; width: 100%;"
        >
          <div
            class="task-item p-4 mb-4 border-2 rounded-lg bg-gray-800 border-green-500"
            data-testid="task-item"
          >
            <div
              class="task-item-header flex justify-between items-center mb-2"
            >
              <h3
                class="task-item-title text-lg font-semibold text-white"
              >
                Completed Task
              </h3>
              <span
                aria-label="completed"
                class="task-item-status px-2 py-1 rounded bg-green-900 text-green-100"
                role="status"
              >
                COMPLETED
              </span>
            </div>
            <p
              class="task-item-description text-gray-300 mb-2"
            >
              Description 3
            </p>
            <p
              class="task-item-due-date text-sm text-gray-400"
            >
              Due: 
              Dec 31, 2024
            </p>
            <div
              class="task-item-controls flex justify-between mt-4"
            >
              <select
                class="task-status-select p-2 border rounded bg-gray-700 text-white border-gray-600 focus:border-blue-500 focus:ring-blue-500"
              >
                <option
                  value="PENDING"
                >
                  Pending
                </option>
                <option
                  value="IN_PROGRESS"
                >
                  In Progress
                </option>
                <option
                  value="COMPLETED"
                >
                  Completed
                </option>
              </select>
              <div
                class="task-item-buttons space-x-2"
              >
                <button
                  class="task-edit-button px-3 py-1 bg-blue-600 text-white rounded hover:bg-blue-700"
                >
                  Edit
                </button>
                <button
                  class="task-delete-button px-3 py-1 bg-red-600 text-white rounded hover:bg-red-600"
                >
                  Delete
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TaskList Snapshots matches snapshot with no tasks 1`] = `
<div>
  <div
    class="task-list-container"
  >
    <div
      class="task-list-filters mb-6 flex flex-col sm:flex-row gap-4"
    >
      <div
        class="flex-1"
      >
        <label
          class="sr-only"
          for="task-search"
        >
          Search tasks
        </label>
        <input
          class="task-search-input w-full px-4 py-2 bg-gray-800 border border-gray-700 rounded-lg text-white placeholder-gray-400"
          id="task-search"
          placeholder="Search tasks..."
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          class="sr-only"
          for="status-filter"
        >
          Filter by status
        </label>
        <select
          aria-label="Filter by status"
          class="task-status-filter px-4 py-2 bg-gray-800 border border-gray-700 rounded-lg text-white"
          id="status-filter"
        >
          <option
            value="ALL"
          >
            All Tasks
          </option>
          <option
            value="PENDING"
          >
            Pending
          </option>
          <option
            value="IN_PROGRESS"
          >
            In Progress
          </option>
          <option
            value="COMPLETED"
          >
            Completed
          </option>
        </select>
      </div>
    </div>
    <div
      class="task-list-empty text-center py-8 text-gray-400"
    >
      No tasks yet. Add your first task!
    </div>
  </div>
</div>
`;
